k8s/services/argocd/components/application.cue

```diff
@@ -31,36 +34,94 @@ import (
 			syncOptions: [...string] | *["CreateNamespace=true"]
 		}
-	}
-	k8sResources: "argocd-\(application.name)": resources.#Application & {
-		metadata: {
-			name:      application.name
-			namespace: k8s.namespace
+		credentials: {
+			privateKey?: v1.#Secret
+			externalSecret: {
+				refreshInterval: *"1h" | string
+				storeRef: {
+					name: string
+					kind: *"ClusterSecretStore" | "SecretStore"
+				}
+				decodingStrategy: *"None" | "Base64" | "Base64URL" | "Auto"
+			}
 		}
-		spec: {
-			destination: {
-				namespace: application.destination.namespace
-				server:    application.destination.server
+	}
+
+	convertGitHubURLToSSH: {
+		input:          application.source.repoURL
+		protocol:       strings.TrimPrefix(input, "https://")
+		withoutSlashes: strings.Replace(protocol, "/", ":", 1)
+		gitURL:         "git@" + withoutSlashes
+		withoutCom:     strings.Replace(gitURL, ".com/", ".com:", 1)
+		output:         withoutCom
+	}
+
+	k8sResources: {
+		"argocd-\(application.name)": resources.#Application & {
+			metadata: {
+				name:      application.name
+				namespace: k8s.namespace
 			}
-			project: application.project
-			source: {
-				repoURL: application.source.repoURL
-				if application.source.path != _|_ {
-					path: application.source.path
+			spec: {
+				destination: {
+					namespace: application.destination.namespace
+					server:    application.destination.server
 				}
-				if application.source.targetRevision != _|_ {
-					targetRevision: application.source.targetRevision
+				project: application.project
+				source: {
+					repoURL: application.source.repoURL
+					if application.source.path != _|_ {
+						path: application.source.path
+					}
+					if application.source.targetRevision != _|_ {
+						targetRevision: application.source.targetRevision
+					}
+					if application.source.chart != _|_ {
+						chart: application.source.chart
+					}
 				}
-				if application.source.chart != _|_ {
-					chart: application.source.chart
+				syncPolicy: {
+					automated: {
+						prune:    application.syncPolicy.automated.prune
+						selfHeal: application.syncPolicy.automated.selfHeal
+					}
+					syncOptions: application.syncPolicy.syncOptions
 				}
 			}
-			syncPolicy: {
-				automated: {
-					prune:    application.syncPolicy.automated.prune
-					selfHeal: application.syncPolicy.automated.selfHeal
+		}
+		if application.credentials.privateKey != _|_ {
+			"argocd-\(application.name)-external-secret": eso.#ExternalSecret & {
+				metadata: {
+					name:      "\(application.name)-repo-secret"
+					namespace: k8s.namespace
+				}
+				spec: {
+					refreshInterval: application.credentials.externalSecret.refreshInterval
+					secretStoreRef: {
+						name: application.credentials.externalSecret.storeRef.name
+						kind: application.credentials.externalSecret.storeRef.kind
+					}
+					template: {
+						stringData: {
+							name: application.name
+							type: "git"
+							url:  convertGitHubURLToSSH.output
```

Comment:
Am I using this function output correctly?

Reply:
we should not need this function